### PR TITLE
added flyweight pattern

### PR DESCRIPTION
**Pull request by janukobytsch**

### CODE REVIEW

File: Structural/Flyweight/CharacterFlyweight.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	 * Clients supply the context-dependent information that the flyweight needs to draw itself
+	 * For flyweights representing characters, extrinsic state usually contains e.g. the font
+	 * @param string $font
+	 */
+	public function draw($font)
+	{
+		print_r("Character {$this->name} printed $font \n");
+	}
+
+}
```
Comment (malukenho) [inline Structural/Flyweight/CharacterFlyweight.php]: EOL EOF

File: Structural/Flyweight/FlyweightFactory.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ */
+class FlyweightFactory {
+	
+	/**
+	 * Associative store for flyweight objects
+     * @var Array
+     */
+	private $pool = array();
+
+	/**
```
Comment (malukenho) [inline Structural/Flyweight/FlyweightFactory.php]: Docblocks align

File: Structural/Flyweight/FlyweightFactory.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        if (!array_key_exists((string) $name, $this->pool)) {
+            $this->pool[(string) $name] = new CharacterFlyweight((string) $name);
+        }
+        return $this->pool[(string) $name];
+    }
+
+    public function totalNumber()
+    {
+    	return sizeof($this->pool);
```
Comment (malukenho) [inline Structural/Flyweight/FlyweightFactory.php]: Do not use tab

File: Structural/Flyweight/FlyweightFactory.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+}
```
Comment (malukenho) [inline Structural/Flyweight/FlyweightFactory.php]: EOL EOF

File: Structural/Flyweight/FlyweightInterface.php [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace DesignPatterns\Structural\Flyweight;
+
+/**
+ * An interface through which flyweights can receive and act on extrinsic state
+ */
+interface FlyweightInterface {
```
Comment (malukenho) [inline Structural/Flyweight/FlyweightInterface.php]: psr

File: Structural/Flyweight/FlyweightInterface.php [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+	public function draw($extrinsicState);
+
+}
```
Comment (malukenho) [inline Structural/Flyweight/FlyweightInterface.php]: EOL EOF

File: Structural/Flyweight/CharacterFlyweight.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace DesignPatterns\Structural\Flyweight;
+
+/**
+ * Implements the flyweight interface and adds storage for intrinsic state, if any.
+ * Instances of concrete flyweights are shared by means of a factory.
+ */
+class CharacterFlyweight implements FlyweightInterface {
```
Comment (malukenho) [inline Structural/Flyweight/CharacterFlyweight.php]: psr

File: Structural/Flyweight/FlyweightInterface.php (hunk 5 above)
Comment (janukobytsch) [inline Structural/Flyweight/FlyweightInterface.php]: can you elaborate on what you mean by psr?

Comment (malukenho) [inline Structural/Flyweight/FlyweightInterface.php]: @janukobytsch code style, see [psr here](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md).

Comment (aik099) [inline Structural/Flyweight/FlyweightInterface.php]: I guess opening brace in class/interface/trait declaration should go on new line.

File: Structural/Flyweight/FlyweightFactory.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	private $pool = array();
+
+	/**
+     * Magic getter
+     * @param string $name
+     * @return Flyweight
+     */
+    public function __get($name)
+    {
+        if (!array_key_exists((string) $name, $this->pool)) {
```
Comment (malukenho) [inline Structural/Flyweight/FlyweightFactory.php]: Do not typecast for string, `__get` magic function ever receive a string.
`public mixed __get(string $name)`

**Comment by janukobytsch** [comment]
Any further remarks?

File: Structural/Flyweight/CharacterFlyweight.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+     * For flyweights representing characters, this is usually the corresponding character code.
+     * @var string
+     */
+    private $name;
+
+    /**
+     * Constructor.
+     * @param string $name
+     */
+    public function __construct($name) {
```
Comment (domnikl) [inline Structural/Flyweight/CharacterFlyweight.php]: did you run [PHP CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer)? I think there are some coding style issues here.

File: Structural/Flyweight/FlyweightInterface.php [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace DesignPatterns\Structural\Flyweight;
+
+/**
+ * An interface through which flyweights can receive and act on extrinsic state
+ */
+interface FlyweightInterface
+{
+    public function draw($extrinsicState);
```
Comment (eddiejaoude) [inline Structural/Flyweight/FlyweightInterface.php]: Need a Docbloc here with **type**

File: Structural/Flyweight/Tests/FlyweightTest.php [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+use DesignPatterns\Structural\Flyweight\FlyweightFactory;
+
+/**
+ * FlyweightTest demonstrates how a client would use the flyweight structure
+ * You don't have to change the code of your client
+ */
+class FlyweightTest extends \PHPUnit_Framework_TestCase
+{
+    private $characters = array('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k',
```
Comment (eddiejaoude) [inline Structural/Flyweight/Tests/FlyweightTest.php]: Docblock here too

**Comment by HassanAlthaf** [comment]
@domnikl I believe this is ready to be merged?

**Comment by domnikl** [comment]
@HassanAlthaf @janukobytsch there are still some coding style fixes to be done here, have a look at @eddiejaoude and my comments in this PR

**Comment by domnikl** [comment]
@janukobytsch would be happy to merge this in if you could fix the remaining coding style fixes?

**Comment by domnikl** [comment]
Fixed CS issues, added README.rst and merged. Thanks!

**Pull request merged**